src/Microsoft.Identity.Client/Exceptions/MsalExceptionFactory.cs

```diff
@@ -52,66 +54,54 @@ public static Exception GetClientException(string errorCode, string errorMessage
+            MsalServiceException exception = isUiRequiredException ?
+                new MsalUiRequiredException(errorCode, errorMessage, innerException) :
+                new MsalServiceException(errorCode, errorMessage, innerException);
+
+            // In most cases we can deserialize the body to get more details such as the suberror
+            OAuth2ResponseBase oAuth2Response =
+                JsonHelper.TryToDeserializeFromJson<OAuth2ResponseBase>(httpResponse?.Body);
+
+            UpdateException(httpResponse, oAuth2Response, exception);
+
+            return exception;
         }
 
-        /// <summary>
-        ///     Throw an <see cref="MsalServiceException" /> exception
-        /// </summary>
         public static Exception GetServiceException(
             string errorCode,
             string errorMessage,
-            Exception innerException,
-            ExceptionDetail exceptionDetail)
+            OAuth2ResponseBase oAuth2Response
+            )
         {
             ValidateRequiredArgs(errorCode, errorMessage);
 
-            return new MsalServiceException(errorCode, errorMessage, innerException)
-            {
-                Claims = exceptionDetail?.Claims,
-                ResponseBody = exceptionDetail?.ResponseBody,
-                StatusCode = exceptionDetail?.StatusCode ?? 0,
-                Headers = exceptionDetail?.HttpResponseHeaders
-            };
+            MsalServiceException exception = new MsalServiceException(errorCode, errorMessage);
+            UpdateException(null, oAuth2Response, exception);
+
+            return exception;
         }
 
-        /// <summary>
-        ///     Throw an <see cref="MsalUiRequiredException" />
-        /// </summary>
-        public static Exception GetUiRequiredException(
-            string errorCode,
-            string errorMessage,
-            Exception innerException,
-            ExceptionDetail exceptionDetail)
+        private static void UpdateException(HttpResponse httpResponse, OAuth2ResponseBase oAuth2Response, MsalServiceException exception)
         {
-            ValidateRequiredArgs(errorCode, errorMessage);
+            exception.ResponseBody = httpResponse?.Body;
+            exception.StatusCode = httpResponse != null ? (int)httpResponse.StatusCode : 0;
+            exception.Headers = httpResponse?.Headers;
 
-            return new MsalUiRequiredException(errorCode, errorMessage, innerException)
-            {
-                Claims = exceptionDetail?.Claims,
-                ResponseBody = exceptionDetail?.ResponseBody,
-                StatusCode = exceptionDetail?.StatusCode ?? 0
-            };
+            exception.Claims = oAuth2Response?.Claims;
+            exception.SubError = oAuth2Response?.SubError;
+            exception.CorrelationId = oAuth2Response?.CorrelationId;
         }
 
         public static string GetPiiScrubbedDetails(Exception ex)
```

Comment:
Why do we have 2 methods which does the same thing?

Reply:
Probably leftover from the ADAL split. Fixed.